src/Datastore/Query/Query.php

```diff
@@ -0,0 +1,376 @@
+    {
+        $this->projectId = $projectId;
+        $this->options = $options + [
+            'query' => [
+                'projection' => [],
+                'kind' => [],
+                'order' => [],
+                'distinctOn' => []
+            ]
+        ];
+    }
+
+    /**
+     * Indicate that this type does support automatic pagination.
+     *
+     * @access private
+     * @return bool
+     */
+    public function canPaginate()
+    {
+        return true;
+    }
+
+    /**
+     * Return a service-compliant array.
+     *
+     * This method is intended for use internally by the PHP client.
+     *
+     * @access private
+     * @return array
+     */
+    public function queryObject()
+    {
+        return array_filter($this->options['query']);
+    }
+
+    /**
+     * Return the query_type union field name.
+     *
+     * @return string
+     * @access private
+     */
+    public function queryKey()
+    {
+        return "query";
+    }
+
+    /**
+     * @access private
+     */
+    public function jsonSerialize()
+    {
+        return $this->queryObject();
+    }
+
+    /**
+     * Set the Query Projection.
+     *
+     * Accepts an array of properties. If set, only these properties will be
+     * returned.
+     *
+     * @param array|string $properties The property or properties to include in
+     *        the result.
+     * @return Query
+     */
+    public function projection($properties)
+    {
+        if (!is_array($properties)) {
+            $properties = [$properties];
+        }
+
+        $this->options['query']['projection'] = $properties;
+
+        return $this;
+    }
+
+    /**
+     * Set the Kind to query.
+     *
+     * If empty, returns entities of all kinds. Must be set in order to filter
+     * results. While you may supply as many kinds as you wish, datastore currently
+     * only accepts one at a time.
+     *
+     * @param array|string $kinds The kind or kinds to return. Only a single kind
+     *        is currently supported.
+     * @return Query
+     */
+    public function kind($kinds)
+    {
+        if (!is_array($kinds)) {
+            $kinds = [$kinds];
+        }
+
+        foreach ($kinds as $kind) {
+            $this->options['query']['kind'][] = $this->propertyName($kind);
+        }
+
+        return $this;
+    }
+
+    /**
+     * Add a filter to the query.
+     *
+     * If the top-level filter is specified as a propertyFilter, it will be replaced.
+     * Any composite filters will be preserved and the new filter will be added.
+     *
+     * @see https://cloud.google.com/datastore/reference/rest/v1/projects/runQuery#operator_1 Allowed Operators
+     *
+     * @param string $property The property to filter.
+     * @param mixed $value The value to check.
+     * @param string $operator The operator to use in the filter.
+     * @return Query
+     */
+    public function filter($property, $value, $operator = self::OP_DEFAULT)
```

Comment:
The order of the params here seems odd to me. I understand it was done so we could have the default `=` operator, but I think I would almost prefer we just require an operator to always be provided if it means it reads better.